### PR TITLE
wdio-cucumber-framework: fix hook handling

### LINKED ISSUE #4175: CucumberJS: `before` and `after` hook are printed to the console as `undefined`
**Environment (please complete the following information):**
 - **WebdriverIO version:** `5.11.1`
 - **Mode:** Testrunner
 - **If WDIO Testrunner, running sync/async:** Sync
 - **Node.js version:** `v10.15.2`
 - **NPM version:** `6.9.0`
 - **Browser name and version:** -
 - **Platform name and version:** -
 - **Additional wdio packages used (if applicable):** `@wdio/spec-reporter`, `@wdio/cucumber-framework`

**Config of WebdriverIO**
An example of how you configured WebdriverIO config

**Describe the bug**
The `before` and `after`-hook are printed in the reporter, see below, but they should be hidden because they don't have a step description

**To Reproduce**
Use the V5 version of WDIO and the latest version of `@wdio/cucumber-framework` (`5.11.1`) and create a simple feature file, add a `before` and` after` hook like this

```js
import { After, Before, Given, Then, When } from 'cucumber';

Before(() => {
  // do nothing
});
After(() => {
  // do nothing
});

Given(/I'm a given step/, () => {});
When(/I'm a when step/, () => {});
Then(/I'm a scenario then step/, () => {});
```

**Expected behavior**
`before` and `after`-hook are not printed

**Log**

```shell
[chrome 75.0.3770.100 Mac OS X #0-0] Spec: /Users/wimselles/Sauce/Git/webdriverio-cucumberjs/__tests__/features/passed.feature
[chrome 75.0.3770.100 Mac OS X #0-0] Running: chrome (v75.0.3770.100) on Mac OS X
[chrome 75.0.3770.100 Mac OS X #0-0]
[chrome 75.0.3770.100 Mac OS X #0-0] Create passed feature
[chrome 75.0.3770.100 Mac OS X #0-0]     Open website
[chrome 75.0.3770.100 Mac OS X #0-0]        ? HookUndefined Step
[chrome 75.0.3770.100 Mac OS X #0-0]        ✓ Given I open "http://webdriver.io/"
[chrome 75.0.3770.100 Mac OS X #0-0]        ✓ Then the title would say "WebdriverIO · Next-gen WebDriver test framework for Node.js"
[chrome 75.0.3770.100 Mac OS X #0-0]        ? HookUndefined Step
[chrome 75.0.3770.100 Mac OS X #0-0]
[chrome 75.0.3770.100 Mac OS X #0-0] 2 passing (2.4s)

Spec Files:      1 passed, 1 total (100% completed) in 00:00:04 
```

### PR DESCRIPTION
**Pull request by mgrybyk**
## Proposed changes

Hook events were emitted like test events that were causing `? HookUndefined Step` to appear.

Now before/after each hooks work like this:
Before hook failure is treated as a test failure.
image

After hook failure is added to report as Hook. Sorry to say there is no information on hook type in event payload, maybe there is someway to improve it.
image

Passed before/after each hooks are not showing up in report.

fix #4175

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

We don't have listeners for BeforeAll and AfterAll at all, so, in case of success results are not shown, failure are treated as worker errors. These hooks are not a part of this PR and should be handled separately later on.
```
[0-0] BeforeAll
0-0 worker error { name: 'VError',
  message:
   'a BeforeAll hook errored, process exiting: test/step-definitions/given.ts:6: beforeAll error',
  stack:
   'VError: a BeforeAll hook errored, process exiting: test/step-definitions/given.ts:6: beforeAll error\n    at /e2e/node_modules/cucumber/lib/runtime/index.js:68:19\n    at Generator.next (<anonymous>)\n    at asyncGeneratorStep (/e2e/node_modules/cucumber/lib/runtime/index.js:26:103)\n    at _next (/e2e/node_modules/cucumber/lib/runtime/index.js:28:194)' }
[0-0] FAILED in chrome - /test/specs/login-success.feature
```

### CODE REVIEW

File: packages/wdio-cucumber-framework/src/utils.js [hunk 1]
```diff
@@ -92,37 +97,44 @@ export function formatMessage ({ type, payload = {} }) {
...
+ * @param {string} type `Step` or `Hook`
+ */
+export function getStepType(type) {
+    return type === 'Step' ? 'test' : 'hook'
+}
+
+/**
+ * build payload for test/hook event
+ */
+export function buildStepPayload(uri, feature, scenario, step, type, title, state, error, duration, passed) {
```
Comment (wswebcreation) [inline packages/wdio-cucumber-framework/src/utils.js]: Is there a way to limit the amount of args and provide an object?

**Comment by mgrybyk** [comment]
upd: added missing test

File: packages/wdio-cucumber-framework/src/reporter.js [hunk 2]
```diff
@@ -54,52 +54,51 @@ class CucumberReporter {
 
     handleBeforeStep (uri, feature, scenario, step, /*sourceLocation*/) {
         this.testStart = new Date()
+        const type = getStepType(step.type)
+        const payload = buildStepPayload(uri, feature, scenario, step, type, getTestStepTitle(step.keyword, step.text))
```
Comment (wswebcreation) [inline packages/wdio-cucumber-framework/src/reporter.js]: This makes longs code lines, see also my comment about the args

File: packages/wdio-cucumber-framework/src/reporter.js [hunk 3]
```diff
@@ -54,52 +54,51 @@ class CucumberReporter {
...
+        const type = getStepType(step.type)
+
+        if (type === 'hook') {
+            return this.afterHook(uri, feature, scenario, step, result)
+        }
+        return this.afterTest(uri, feature, scenario, step, result)
+    }
+
+    afterHook (uri, feature, scenario, step, result) {
+        const payload = buildStepPayload(uri, feature, scenario, step, 'hook', getTestStepTitle(step.keyword, step.text, 'hook'), result.status, result.exception, new Date() - this.testStart)
```
Comment (wswebcreation) [inline packages/wdio-cucumber-framework/src/reporter.js]: This makes longs code lines, see also my comment about the args

File: packages/wdio-cucumber-framework/src/utils.js [hunk 4]
```diff
@@ -92,37 +97,44 @@ export function formatMessage ({ type, payload = {} }) {
         type: type
     }
 
-    if (payload.error) {
```
Comment (mgrybyk) [inline packages/wdio-cucumber-framework/src/utils.js]: this one is mocha specific

File: packages/wdio-cucumber-framework/src/utils.js [hunk 5]
```diff
@@ -92,37 +97,44 @@ export function formatMessage ({ type, payload = {} }) {
...
-        if (payload.title && payload.title.match(/^"(before|after)( all)*" hook/g)) {
-            message.type = 'hook:end'
-        }
-    }
-
-    /**
-     * Add the current test title to the payload for cases where it helps to
-     * identify the test, e.g. when running inside a beforeEach hook
-     */
-    if (payload.ctx && payload.ctx.currentTest) {
```
Comment (mgrybyk) [inline packages/wdio-cucumber-framework/src/utils.js]: also mocha specific

File: packages/wdio-cucumber-framework/src/utils.js [hunk 6]
```diff
@@ -92,37 +97,44 @@ export function formatMessage ({ type, payload = {} }) {
...
-        message.currentTest = payload.ctx.currentTest.title
-    }
-
-    if (type.match(/Test/)) {
```
Comment (mgrybyk) [inline packages/wdio-cucumber-framework/src/utils.js]: this condition if always false

File: packages/wdio-cucumber-framework/src/utils.js (hunk 1 above)
Comment (mgrybyk) [reply packages/wdio-cucumber-framework/src/utils.js]: Yeah, I was also thinking to do this and merge object instead

Comment (mgrybyk) [reply packages/wdio-cucumber-framework/src/utils.js]: @wswebcreation done

**Pull request merged**